### PR TITLE
Littleredcomputer/cse

### PR DESCRIPTION
**Pull request by littleredcomputer**

### CODE REVIEW

File: src/emmy/expression/cse.cljc [hunk 1]
```diff
@@ -7,152 +7,154 @@
...
+;; but should be perfectly transparent to the code generators of the host
+;; languages.)
+
+(defn- dissociate
+  "If `x` is an expression like `(f x1 x2 x3...)`, where `(f? f)` is true,
+  returns `(f ... (f x1 x2) x3 ...)`, i.e., we replace applications with
+  arity > 2 with left-associative nested applications of arity 2. Otherwise
+  `x` is returned."
+  [f? x]
+  (letfn [(go [x]
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:61]: Couple of ideas, not sure if they are helpful at all (your call either way!)

```clojure
(defn- dissociate [f? x]
  (letfn [(go [x]
              (cond (not (seq? x)) x
                    
                    ;; I am irrationally scared of laziness in cases like this, after being bitten in the simplifier
                    ;; implementation by a lazy `map`. Maybe it doesn't matter here, but you could `doall`...
                    (or (not (f? (first x)))
                        (< (count x) 4))
                    (doall (map go x))
                    
                    :else
                    ;; transduce would work here too and save an intermediate sequence... wdyt?
                    (let [[f x & xs] x]
                      (transduce (partial list f)
                                 (map go)
                                 x
                                 xs))))]
    (go x)))
```

File: src/emmy/expression/cse.cljc [hunk 2]
```diff
@@ -7,152 +7,154 @@
...
+    (go x)))
+
+;; This is the heart of the F90 algorithm. We do a post-order traversal
+;; of the tree, building an association between equivalence classes of
+;; subtree to small integers. We can then represent the maximally factored
+;; DAG as a simple sequence of expressions, where integers in argument
+;; position represent back-references to the indexed computation emitted
+;; previously (that is, in the nth expression, indices in the range [0..n-1]
+;; may appear). Thus the computation may be efficiently performed by
+;; evaluation the subexpressions in the order reported.
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:77]: "by evaluating"

File: src/emmy/expression/cse.cljc [hunk 3]
```diff
@@ -7,152 +7,154 @@
...
+              (or (@table x)
+                  (let [i (swap! counter inc)]
+                    (swap! table assoc x i)
+                    i)))
+            (uid [x]
+                (cond
+                  (symbol? x) x
+                  (number? x) (install x)
+                  (seq? x) (install (into [(first x)] (map uid (next x))))
+                  :else (throw (u/exception (str "cse? " x)))))]
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:105]: I think `u/exception` already throws, no need for the `throw` wrapper

File: src/emmy/expression/cse.cljc [hunk 4]
```diff
@@ -7,152 +7,154 @@
...
+  by new expressions; in this way, multiple expressions in the same scope
+  may be compacted.
+
+  Passing the argument `:table` will result in a sequence
+  of simple forms with a function symbol at the head followed by arguments.
+  The arguments are either symbols drawn from the original expression or
+  integers which represent the the index of the value of a computation
+  performed earlier in the sequence."
+  []
+  (let [counter (atom -1)
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:93]: could use a `volatile!` here, `(volatile! -1)`, and then `vswap!` below...

From https://github.com/clojure/clojure/blob/97c1210713cbbe0bacbb59004a91e25515d74ab6/changes.md#21-transducers :

> volatiles - there are a new set of functions (volatile!, vswap!, vreset!, volatile?) to create and use volatile "boxes" to hold state in stateful transducers. Volatiles are faster than atoms but give up atomicity guarantees so should only be used with thread isolation.

File: src/emmy/expression/cse.cljc [hunk 5]
```diff
@@ -7,152 +7,154 @@
...
+        table   (atom {})]
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:94]: A volatile would be faster here too, see the distinct transducer for a similar spot that uses this: https://github.com/clojure/clojure/blob/master/src/clj/clojure/core.clj#L5075

File: src/emmy/expression/cse.cljc [hunk 6]
```diff
@@ -7,152 +7,154 @@
...
+    (letfn [(install [x]
+              (or (@table x)
+                  (let [i (swap! counter inc)]
+                    (swap! table assoc x i)
+                    i)))
+            (uid [x]
+                (cond
+                  (symbol? x) x
+                  (number? x) (install x)
+                  (seq? x) (install (into [(first x)] (map uid (next x))))
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:104]: Nitpick, prefer the transducer form of 

```clojure
(seq? x) (install (into [(first x)] (map uid) (next x)))
```
just one paren moved!

File: src/emmy/expression/cse.cljc [hunk 7]
```diff
@@ -7,152 +7,154 @@
...
+
+(defn- uid-assigner
+  "Produces a function which takes an S-expression and produces the
+  compacted DAG representation of the tree, as described in [Flajolet 90].
+  Returns the ID of the root of the tree. The assigner generated by this
+  function can be used multiple times, and earlier compact forms will be reused
+  by new expressions; in this way, multiple expressions in the same scope
+  may be compacted.
+
+  Passing the argument `:table` will result in a sequence
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:87]: if you don't want the magic keyword, you could trigger this behavior with a no-arg call?

File: src/emmy/expression/cse.cljc [hunk 8]
```diff
@@ -7,152 +7,154 @@
...
+                    i)))
+            (uid [x]
+                (cond
+                  (symbol? x) x
+                  (number? x) (install x)
+                  (seq? x) (install (into [(first x)] (map uid (next x))))
+                  :else (throw (u/exception (str "cse? " x)))))]
+      (fn [expr]
+        (if (= expr :table)
+          (->> @table (sort-by second) (map first))
```
Comment (sritchie) [inline src/emmy/expression/cse.cljc:108]: this is another case where I will recommend a `doall`, given that this is querying a stateful data structure and we probably want the query to trigger at the time of the call.

**Review by sritchie — COMMENTED** [review]
Looks really good, and I have been playing with the results over at emmy-viewers. I want to stare a bit more at the algorithm tomorrow but here are my first responses...

File: src/emmy/expression/compile.cljc [hunk 9]
```diff
@@ -341,37 +341,24 @@
...
+  (letfn [(children? [x]
+            (and (sequential? x) ('#{up down vector} (first x))))
+          (aset-form [index value]
+            `(aset ~index ~value))]
+    (if (= calling-convention :primitive)
+      (update code :body
+              (fn [body]
+                (if (children? body)
+                  (let [array-symbol (nth argv 1)
+                        values (filter (complement children?) (tree-seq children? rest body))]
```
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc:359]: done.

File: test/emmy/expression/compile_test.cljc [hunk 10]
```diff
@@ -130,29 +154,53 @@
 
 
           (doseq [[mode expected-source]
-                  {:clj `(fn ~'[[y1 [y2 [y3 [y4]]]]]
-                           (vector (+ (* 3.1 ~'y1) 1.55)))
-                   :js ["[y1, [y2, [y3, [y4]]]]" "  return [3.1 * y1 + 1.55];"]}]
+                  {:clj '(fn [[y1 [y2 [y3 [y4]]]]]
+                           (let [_5 3.1
+                                              _6 (* _5 y1)
```
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply test/emmy/expression/compile_test.cljc:159]: ✅

File: src/emmy/expression/cse.cljc (hunk 8 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:108]: But the value of `@table` is immutable?

File: src/emmy/expression/cse.cljc (hunk 7 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:87]: ✅

File: src/emmy/expression/cse.cljc (hunk 8 above)
Comment (sritchie) [reply src/emmy/expression/cse.cljc:108]: oh, yeah, good call! Told you I had paranoia.

File: src/emmy/expression/cse.cljc (hunk 6 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:104]: thank you. I have to get better at using transducers.

File: src/emmy/expression/cse.cljc (hunk 5 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:94]: We have the thread isolation that is called for, but my own preference would be to reserve volatile for situations where the mutable thing is created and disposed of within the function application itself even though this is unnecessarily conservative as you point out.

File: src/emmy/expression/cse.cljc (hunk 4 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:93]: see below

File: src/emmy/expression/cse.cljc (hunk 3 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:105]: ✅

File: src/emmy/expression/cse.cljc (hunk 2 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:77]: ✅

File: src/emmy/expression/cse.cljc (hunk 1 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:61]: I think it is reasonable for this function to be lazy, which would be useful if this were one in a sequence of tree-transformations, and that might happen as the compiler gets more powerful. Looking at where the laziness is capped in this case, it's an `into []` enforced by its caller.

`transduce` hint gratefully accepted!

Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:61]:
```clojure
(transduce (map str) (partial list 'f) 'x [2 3 4])
(f (f (f (f x "2") "3") "4"))
```
Annoyingly an arity-1 call to f is used to finalize the reduction (grr). I think reduce is more transparent here

File: src/emmy/expression/compile.cljc (hunk 9 above)
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc:359]: ✅

File: src/emmy/expression/cse.cljc (hunk 1 above)
Comment (littleredcomputer) [reply src/emmy/expression/cse.cljc:61]: @sritchie points out that `completing` would fix that problem, but I've stuck with reduce on the KISS principle

**Pull request merged**